### PR TITLE
feat: Cal.com Copilot

### LINKED ISSUE #13341: [CAL-2982] Cal.com Copilot
Implemented in [https://github.com/calcom/cal.com/pull/12855](https://github.com/calcom/cal.com/pull/12855)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2982](https://linear.app/calcom/issue/CAL-2982/calcom-copilot)</sub>

### PR DESCRIPTION
**Pull request by maxprilutskiy**
Fixes #13341 

Hey @keithwillcode @baileypumfleet,

As agreed, submitting a PR with what we have done during the hackathon for you to take a look. :)

P.S. Commiting the work as-is: likely has redundancies/inefficiencies/uncovered edge cases (since it's a hackathon PoC).

# Cal.com Copilot

Cal.com Copilot helps new users get started with the platform faster and provides a much more personalized first experience, using AI.

Built by [@vrcprl](https://github.com/vrcprl) and [@maxprilutskiy](https://github.com/maxprilutskiy), during the [#OSSHack](https://osshack.com) hackathon in NYC.

---

Our solution improves the onboarding flow and creates a personalized set of events for users, offering more choices to schedule new calls.

Previously, the app randomly generated three events at the end of the user onboarding process. Now, based on the information provided by users during onboarding, the app generates several personalized events.

These include:

* Title
* Description
* Location (in-person, by phone, online), along with specific details such as phone numbers or addresses if provided
* Meeting durations depend on the nature of the meeting (for example, a dinner with friends might last about 2 hours, while a quick chat usually doesn't take more than 15 minutes)
* Minimum notice requirements; if the user specifies different notice times for various event types (e.g., 2 days for in-person and 2 hours for online), the app will configure accordingly
* Automatically identifying personal events (like a call with mom) and hiding them from the list of events visible to business partners

## Demo

Check out our demo 😉

[Cal.com Copilot Demo Loom](https://www.loom.com/share/3ca21a10254443b9a01bdc90e78d3a5d)

<https://www.loom.com/share/3ca21a10254443b9a01bdc90e78d3a5d>

## What's next

Opportunities for further improvements:

* Add support for more features (settings) that can be configured based on user input;
* Use this solution to create a single event based on the user’s input (aka use co-pilot outside of the onboarding flow);
* Make Copilot configurable, e.g. turn on/off depending on the presence of the openai key;
* Migrate from openai api to another solution: openai can sometimes be unstable, but there are other alternatives out there;
* Improve not only the actual loading speed, but also the perceived loading speed via specific design decisions;
* Add user input validation;
* Add needed translations;
* Make the prompts respect the zod schema, by either fine-tuning the LLM model or providing an extensive set of input <> output examples within the LLM prompts themselves;
* Make the prompts much more robust by testing them against hundreds of potential user inputs;
* Implement a JSON validator, as there will never be 100% certainty openai returns a valid JSON;
* Implement fallback mechanism if OpenAI returns invalid response or is down;
* Ideally, the feature must also be connected to product analytics tool you're using (posthog?), to make sure we arent' releasing this blindly.

## Rollout considerations

We recommend rolling this out in phases (e.g. 20 => 60 => 100), to make sure Copilot not just gives new users more reasons to be booked, but those bookings **actually happen** and the variety of new event types **doesn't overwhelm** the user.

### CODE REVIEW

File: packages/trpc/server/routers/viewer/eventTypes/createCopilotSuggestion.handler.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      {
+        role: "assistant",
+        content: `Sure! Here's the array, in valid JSON format:`,
+      },
+    ],
+  });
+
+  let responseContent = llmResponse?.choices?.[0]?.message?.content || "[]";
+
+  console.log("responseContent", responseContent);
```
Comment (exception) [inline packages/trpc/server/routers/viewer/eventTypes/createCopilotSuggestion.handler.ts:136]:
```suggestion
```

**Comment by exception** [comment]
For this PR to move forward, we'd need the option for AI Copilot or the default event types. 
I believe this should also be done in the background (or at least have a notice) as it slows the onboarding process (while AI compute is happening).

cc @keithwillcode @PeerRich

**Comment by keithwillcode** [comment]
Moving back to draft as there are implementation changes we are going to make

**Comment by keithwillcode** [comment]
Closing for now until we re-prioritize this

**Pull request closed without merging**